zerver/lib/message.py

```diff
@@ -554,3 +558,130 @@ def apply_unread_message_event(state, message):
+
+        if len(stream_messages) > 0:
+            # We do this check for last event since if the last subscription
+            # event was a subscription_deactivated then we don't want to create
+            # UserMessage rows for any of the remaining messages.
+            if stream_subscription_logs[-1].event_type in (
+                    'subscription_activated',
+                    'subscription_created'):
+                for stream_message in stream_messages:
+                    store_user_message_to_insert(stream_message)
+
+def add_missing_messages(user_profile):
+    # type: (UserProfile) -> None
+    # This list will store all the messages for which we eventually will create
+    # UserMessage table rows by doing a bulk insert.
+    all_messages_to_insert = []  # type: List[UserMessage]
+
+    all_stream_subs = list(Subscription.objects.select_related('recipient').filter(
+        user_profile=user_profile,
+        recipient__type=Recipient.STREAM).values('recipient', 'recipient__type_id'))
+
+    # For Stream messages we need to check messages against data from
+    # RealmAuditLog for visibility to user. So we fetch the subscription logs.
+    stream_ids = [sub['recipient__type_id'] for sub in all_stream_subs]
+    events = ['subscription_created', 'subscription_deactivated', 'subscription_activated']
+    subscription_logs = list(RealmAuditLog.objects.select_related(
+        'modified_stream').filter(
+        modified_user=user_profile,
+        modified_stream__id__in=stream_ids,
+        event_type__in=events).order_by('event_last_message_id'))
+    all_stream_subscription_logs = defaultdict(list)  # type: DefaultDict[int, List]
+    for log in subscription_logs:
+        all_stream_subscription_logs[log.modified_stream.id].append(log)
+
+    recipient_ids = []
+    for sub in all_stream_subs:
+        stream_subscription_logs = all_stream_subscription_logs[sub['recipient__type_id']]
+        if (stream_subscription_logs[-1].event_type == 'subscription_deactivated' and
+                stream_subscription_logs[-1].event_last_message_id < user_profile.last_active_message_id):
+            # We are going to short circuit this iteration as its no use
+            # iterating since user unsubscribed before soft-deactivation
+            continue
+        recipient_ids.append(sub['recipient'])
+
+    all_stream_msgs = list(Message.objects.select_related(
+        'recipient').filter(
+        recipient__id__in=recipient_ids,
+        id__gt=user_profile.last_active_message_id).order_by('id').values(
+        'id', 'recipient__type_id'))
+    already_created_ums = list(UserMessage.objects.select_related(
+        'message').filter(
+        user_profile=user_profile,
+        message__recipient__type=Recipient.STREAM,
+        message__id__gt=user_profile.last_active_message_id).values(
+        'message__id'))
+    already_created_ums = [msg['message__id'] for msg in already_created_ums]
```

Comment:
This should use a `set()`, otherwise the next line will have quadratic performance, since the `not in` operator is linear time when accessing a list.